--- pyOpenSSL-0.7/setup.py	2008-04-11 16:53:24.000000000 +0100
+++ pyOpenSSL-0.7/setup-new.py	2008-09-15 23:46:19.000000000 +0100
@@ -64,6 +64,9 @@
 if sys.platform == 'darwin':
     IncludeDirs = ['/sw/include']
     LibraryDirs = ['/sw/lib']
+elif os.name == 'posix':
+    IncludeDirs = ['/usr/kerberos/include']
+    LibraryDirs = ['/usr/kerberos/lib']
 
 def mkExtension(name):
     modname = 'OpenSSL.' + name
